core/pkg/artifacts/saver.go

```diff
@@ -144,99 +153,165 @@ func (as *ArtifactSaver) uploadFiles(artifactID string, manifest *Manifest, mani
-					func(t *filetransfer.Task) {
-						taskResultsChan <- TaskResult{t, name}
-					},
-				)
-				as.FileTransferManager.AddTask(task)
-			}
 		}
-		// Wait for filetransfer to catch up. If there's nothing more to schedule, wait for all in progress tasks.
-		for numInProgress > maxBacklog || (len(fileSpecsBatch) == 0 && numInProgress > 0) {
-			numInProgress--
-			result := <-taskResultsChan
-			if result.Task.Err != nil {
-				// We want to retry when the signed URL expires. However, distinguishing that error from others is not
-				// trivial. As a heuristic, we retry if the request failed more than an hour after we fetched the URL.
-				if time.Since(nameToScheduledTime[result.Name]) < 1*time.Hour {
-					return result.Task.Err
-				}
-				delete(nameToScheduledTime, result.Name) // retry
+		if len(fileSpecs) > 0 {
+			as.Logger.Warn("some files failed to upload, retrying", "count", len(fileSpecs))
+		}
+	}
+	return nil
+}
+
+func (as *ArtifactSaver) processFiles(
+	manifest *Manifest, namedFileSpecs map[string]gql.CreateArtifactFileSpecInput,
+) (map[string]gql.CreateArtifactFileSpecInput, error) {
+	// Channels to get responses from the batch url retrievers.
+	readyChan := make(chan serverFileResponse)
+	errorChan := make(chan error)
+
+	doneChan := make(chan *filetransfer.Task)
+	mustRetry := map[string]gql.CreateArtifactFileSpecInput{}
+
+	numActive := 0
+
+	var batch []gql.CreateArtifactFileSpecInput
+	fileSpecs := []gql.CreateArtifactFileSpecInput{}
+	for _, spec := range namedFileSpecs {
+		fileSpecs = append(fileSpecs, spec)
+	}
+
+	for as.numDone+len(mustRetry) < as.numTotal {
+		// Start new batches until we get to the desired number of active uploads.
+		for len(fileSpecs) > 0 && numActive <= (as.maxActiveBatches-1)*as.batchSize() {
+			batch, fileSpecs = as.nextBatch(fileSpecs)
+			numActive += len(batch)
+			go as.batchFileDataRetriever(batch, readyChan, errorChan)
+		}
+		select {
+		// Start any uploads that are ready.
+		case fileInfo := <-readyChan:
+			entry := manifest.Contents[fileInfo.Name]
+			entry.BirthArtifactID = &fileInfo.BirthArtifactID
+			manifest.Contents[fileInfo.Name] = entry
+			as.cacheEntry(entry)
+			if fileInfo.UploadUrl == nil {
+				// The server already has this file.
+				numActive--
+				as.numDone++
 				continue
 			}
-			numDone++
-			entry := manifest.Contents[result.Name]
-			if !entry.SkipCache {
-				digest := entry.Digest
-				go func() {
-					err := as.FileCache.AddFileAndCheckDigest(result.Task.Path, digest)
-					if err != nil {
-						slog.Error("error adding file to cache", "err", err)
-					}
-				}()
+			task := newUploadTask(fileInfo, *entry.LocalPath)
+			task.SetCompletionCallback(func(t *filetransfer.Task) { doneChan <- t })
+			as.FileTransferManager.AddTask(task)
+		// Listen for completed uploads, adding to the retry list if they failed.
+		case result := <-doneChan:
+			numActive--
+			if result.Err != nil {
+				mustRetry[result.Name] = namedFileSpecs[result.Name]
```

Comment:
this looks like the only line that relies on `namedFileSpecs`, if we refactor this to use a channel, I'm not sure if you ever need `namedFileSpecs` vs just the list. and that would save some map -> list conversion